### PR TITLE
DOC-221 New Rate-Limiting Article.

### PR DESCRIPTION
**Pull request by janettelynch**

### CODE REVIEW

**Review by plfx — DISMISSED** [review]

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# J1QL Query Rate Limiting
+
+J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query uses one token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API.
+
+#### **Account Rate Limit for J1QL Queries**
+
+The default rate of J1QL queries is`120 tokens per 1-minute period`.
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:7]:
```suggestion
The default rate of J1QL queries is`120 tokens per 2-minute period`.
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+- At any time, there are a maximum of 120 tokens available for J1QL queries.
+- JupiterOne declines queries when the rate exceeds 120 tokens per 1-minute period.
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:10]:
```suggestion
- JupiterOne declines queries when the rate exceeds 120 tokens per 2-minute period.
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+- **Average rate allowed:** 2 J1QL queries per second.
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:11]:
```suggestion
- **Average rate allowed:** 1 J1QL query per second.
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+- **Configurable:** JupiterOne can adjust this rate limit on your behalf. Contact your J1 administrator.
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:12]: I'm not sure who is best to make a determination on this point, but I think we want to tie increase or removal of this limit to some kind of business value.   E.g. do we automatically approve all requests from enterprise customers to up to the next level (has been our typical first upgrade)?  

Maybe we could implement something like the following:

**Default limit**: 120 tokens / 120 seconds
**No questions asked upgrade for enterprise customers**: 60 tokens / 60 seconds
**Request for approval**: 120 tokens / 60 seconds
**Not public, select customers only**: no rate limit

And then we'd need to determine what that request for approval process looks like.

@a-u-h-g

**Review by benrj — CHANGES_REQUESTED** [review]

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# J1QL Query Rate Limiting
+
+J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query uses one token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API.
```
Comment (a-u-h-g) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:3]:
```suggestion
J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query consumes one request token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API.
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md (hunk 3 above)
Comment (a-u-h-g) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:11]: @benrj is this correct? avg rate allowed is 2 queries / sec?

File: knowledgeBase/APIs/rate-limiting-j1ql.md (hunk 4 above)
Comment (a-u-h-g) [reply knowledgeBase/APIs/rate-limiting-j1ql.md:12]: Agreed with the above proposal. How do you think we should change our current public facing info in sales purchase orders if at all? "30 requests/min = default, 60 requests/min = no question upgrade for enterprise"

**Review by a-u-h-g — COMMENTED** [review]
Some comments

File: knowledgeBase/APIs/rate-limiting-j1ql.md (hunk 3 above)
Comment (benrj) [reply knowledgeBase/APIs/rate-limiting-j1ql.md:11]: It should be 1 query per second at the default level.

File: knowledgeBase/APIs/rate-limiting-j1ql.md (hunk 4 above)
Comment (janettelynch) [reply knowledgeBase/APIs/rate-limiting-j1ql.md:12]: @benrj Can we make it a bit more reader friendly this way:

Default limit: 120 tokens / 120 seconds
Enterprise user upgrade: 60 tokens / 60 seconds
Request for approval: 120 tokens / 60 seconds
Select users: no rate limit


So, I'm replacing lines 9-12 with the above?

Comment (benrj) [reply knowledgeBase/APIs/rate-limiting-j1ql.md:12]: I would suggest replacing just line 12 with:

```
Enterprise customers are eligible for a rate limit increase up to 120 tokens / 60 seconds, subject to usage analysis and approval.  Contact your TSE to request an increase.

All customers should incorporate retry with backoff in response to rate limiting responses from the J1 API.
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# J1QL Query Rate Limiting
+
+J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query consumes one request token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API. You should incorporate retry with backoff in response to rate-limiting responses from the J1 API.
+
+#### **Account Rate Limit for J1QL Queries**
+
+- Default limit: 120 tokens/120 seconds
+- Enterprise user upgrade: 60 tokens/60 seconds
+- Request for approval: 120 tokens/60 seconds
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:9]:
```suggestion
```
I think this is redundant.

File: knowledgeBase/APIs/rate-limiting-j1ql.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# J1QL Query Rate Limiting
+
+J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query consumes one request token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API. You should incorporate retry with backoff in response to rate-limiting responses from the J1 API.
+
+#### **Account Rate Limit for J1QL Queries**
+
+- Default limit: 120 tokens/120 seconds
+- Enterprise user upgrade: 60 tokens/60 seconds
```
Comment (benrj) [inline knowledgeBase/APIs/rate-limiting-j1ql.md:8]:
```suggestion
- Enterprise limit (available on request): 60 tokens/60 seconds
```

File: knowledgeBase/APIs/rate-limiting-j1ql.md (hunk 6 above)
Comment (janettelynch) [reply knowledgeBase/APIs/rate-limiting-j1ql.md:9]: Ok, removing line 9. Thanks!

**Review by benrj — APPROVED** [review]

**Review by sarabeth — APPROVED** [review]

**Pull request merged**